bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java

```diff
@@ -0,0 +1,379 @@
+    private @Nullable AbstractPresentableCalendar runtimeCalendar;
+    private @Nullable ScheduledFuture<?> updateJobFuture;
+    private Instant updateStatesLastCalledTime;
+
+    public ICalendarHandler(Thing thing, HttpClient httpClient, EventPublisher eventPublisher) {
+        super(thing);
+        this.httpClient = httpClient;
+        calendarFile = new File(ConfigConstants.getUserDataFolder() + File.separator
+                + getThing().getUID().getAsString().replaceAll("[<>:\"/\\\\|?*]", "_") + ".ical");
+        eventPublisherCallback = eventPublisher;
+        updateStatesLastCalledTime = Instant.now();
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> currentUpdateJobFuture = updateJobFuture;
+        if (currentUpdateJobFuture != null) {
+            currentUpdateJobFuture.cancel(true);
+        }
+        ScheduledFuture<?> currentPullJobFuture = pullJobFuture;
+        if (currentPullJobFuture != null) {
+            currentPullJobFuture.cancel(true);
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        switch (channelUID.getId()) {
+            case CHANNEL_CURRENT_EVENT_PRESENT:
+            case CHANNEL_CURRENT_EVENT_TITLE:
+            case CHANNEL_CURRENT_EVENT_START:
+            case CHANNEL_CURRENT_EVENT_END:
+            case CHANNEL_NEXT_EVENT_TITLE:
+            case CHANNEL_NEXT_EVENT_START:
+            case CHANNEL_NEXT_EVENT_END:
+                if (command instanceof RefreshType) {
+                    updateStates();
+                }
+                break;
+            default:
+                logger.warn("Framework sent command to unknown channel with id '{}'", channelUID.getId());
+        }
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+
+        ICalendarConfiguration currentConfiguration = getConfigAs(ICalendarConfiguration.class);
+        configuration = currentConfiguration;
+
+        if ((currentConfiguration.username == null && currentConfiguration.password != null)
+                || (currentConfiguration.username != null && currentConfiguration.password == null)) {
+            logger.warn("Only one of username and password was set. This is invalid.");
+            updateStatus(ThingStatus.OFFLINE);
+            return;
+        }
+
+        PullJob regularPull;
+        try {
+            regularPull = new PullJob(httpClient, new URI(currentConfiguration.url), currentConfiguration.username,
+                    currentConfiguration.password, calendarFile, currentConfiguration.maxSize * 1048576, this);
+        } catch (URISyntaxException e) {
+            logger.warn(
+                    "The URI '{}' for downloading the calendar contains syntax errors. This will result in no downloads/updates.",
+                    currentConfiguration.url, e);
+            updateStatus(ThingStatus.OFFLINE);
```

Comment:
Maybe mark this as a configuration error as well?

Reply:
OK